### DISCUSSION TITLE
"Local" plugin for marp-team.marp-vscode?

### DISCUSSION THREAD

**Discussion by thmsklngr** [marp-team/marp#432] [Q&A]
Hi,

I'm using Marp within VSCode and after some time I start to love it. The only thing is that I sometimes miss things like 2 columns on one slide. After some time searching I found a solution here to create some kind of plugin for this but I did not find any answer on how to create a plugin, lets say specific for this one project.
I know that there's a solution for CSS with a local theme.css to _create_ your own theme by extending an existing one, but is it also possible to so for plugins in a similar way? Or is there a general recommendation on how to add self written plugins when using Marp in VSCode?
Many thanks in advance for any answer!
Regards, Thomas

**Comment by yhatt** [comment]
If you are looking for how to create a plugin, please refer to #206.

In Marp ecosystem, plugins brought by the user are always applicable to [Marp CLI](https://github.com/marp-team/marp-cli) only. Marp for VS Code has no ability to extend Marp engine through plugin due to security reason, and currently we have not any plans to allow plugins in VS Code extension.

I have repeated the same description every time someone asked about this.

> Using customized Marp engine is powerful, but with great power comes great responsibility. For example, [Marp extension might be used as an attack vector to steal some important data in your PC if enabled the engine customization by JS.](https://code.visualstudio.com/blogs/2021/07/06/workspace-trust#:~:text=What%20about%20the,to%20be%20owned.)
>
> _&mdash; https://github.com/orgs/marp-team/discussions/164#discussioncomment-1313187_

[A classic Marp](https://yhatt.github.io/marp/) had burned out due to a weakness of security, so we are being careful not to create a hole. Marp for VS Code extension is easy to use for newbies, so we are especially taking care not to allow customization via (malicious) plugins.

We treat users using plugin are advanced users, so we recommend to use Marp CLI, designed for advanced tasks. If you do not want to leave from VS Code, you just have to use Marp CLI in [VS Code integrated terminal](https://code.visualstudio.com/docs/editor/integrated-terminal).
